Mathlib/Combinatorics/SimpleGraph/EdgeLabelling.lean

```diff
@@ -0,0 +1,227 @@
+/-
+Copyright (c) 2023 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta, Olivia Röhrig
+-/
+import Mathlib.Combinatorics.SimpleGraph.Maps
+import Mathlib.Combinatorics.SimpleGraph.Finite
+import Mathlib.Data.Fintype.BigOperators
+import Mathlib.Data.ZMod.Defs
+
+/-!
+# Edge labellings
+
+This module defines labellings of the edges of a graph.
+
+## Main definitions
+
+- `SimpleGraph.EdgeLabelling`: An assignment of a label from a given type to each edge of the graph.
+
+- `SimpleGraph.EdgeLabelling.labelGraph`: the graph consisting of all edges with a given label.
+-/
+
+open Finset
+open Fintype (card)
+
+namespace SimpleGraph
+
+variable {V V' : Type*} {G : SimpleGraph V} {G' : SimpleGraph V'} {K K' : Type*}
+
+/-- An edge labelling of a simple graph `G` with labels in type `K`. Sometimes this is called an
+edge-colouring, but we reserve that terminology for labellings where incident edges cannot share a
+label.
+-/
+def EdgeLabelling (G : SimpleGraph V) (K : Type*) :=
+  G.edgeSet → K
+
+instance [DecidableEq V] [Fintype G.edgeSet] [Fintype K] : Fintype (EdgeLabelling G K) :=
+  Pi.instFintype
+
+instance [Nonempty K] : Nonempty (EdgeLabelling G K) :=
+  Pi.instNonempty
+
+instance [Inhabited K] : Inhabited (EdgeLabelling G K) :=
+  Pi.instInhabited
+
+instance [Subsingleton K] : Subsingleton (EdgeLabelling G K) :=
+  Pi.instSubsingleton
+
+instance [Unique K] : Unique (EdgeLabelling G K) :=
+  Pi.unique
+
+/--
+An edge labelling of the complete graph on `V` with labels in type `K`. Sometimes this is called an
+edge-colouring, but we reserve that terminology for labellings where incident edges cannot share a
+label.
+-/
+abbrev TopEdgeLabelling (V K : Type*) :=
+  EdgeLabelling (⊤ : SimpleGraph V) K
+
+theorem card_topEdgeLabelling [DecidableEq V] [Fintype V] [Fintype K] :
+    card (TopEdgeLabelling V K) = card K ^ (card V).choose 2 :=
+  Fintype.card_fun.trans (by rw [← edgeFinset_card, card_edgeFinset_top_eq_card_choose_two])
+
+namespace EdgeLabelling
+
+/--
+Convenience function to get the colour of the edge `x ~ y` in the colouring of the complete graph
+on `V`.
+-/
+def get (C : EdgeLabelling G K) (x y : V) (h : G.Adj x y) : K :=
+  C ⟨s(x, y), h⟩
+
+lemma get_eq (C : EdgeLabelling G K) (x y : V) (h : G.Adj x y) : C.get x y h = C ⟨s(x, y), h⟩ :=
+  rfl
+
+variable {C : EdgeLabelling G K}
+
+theorem get_comm (x y : V) (h) : C.get y x h = C.get x y h.symm := by
+  simp [EdgeLabelling.get, Sym2.eq_swap]
+
+@[ext]
+theorem ext_get {C' : EdgeLabelling G K}
+    (h : ∀ x y, (h : G.Adj x y) → C.get x y h = C'.get x y h) : C = C' := by
+  refine funext fun ⟨e, he⟩ => ?_
+  induction e using Sym2.inductionOn
+  exact h _ _ he
```

Comment:
```
  refine funext fun ⟨e, he⟩ => ?_
```
This line came from porting before we had `ext`, I'd hope it can be written more simply now?